crates/node/src/txvalidation/mod.rs

```diff
@@ -0,0 +1,203 @@
+pub enum EventProcessError {
+    #[error("Fail to rcv Tx from the channel: {0}")]
+    RcvChannelError(#[from] tokio::sync::oneshot::error::RecvError),
+    #[error("Fail to send the Tx on the channel: {0}")]
+    SendChannelError(
+        #[from]
+        tokio::sync::mpsc::error::SendError<(Transaction<TxReceive>, Option<CallbackSender>)>,
+    ),
+    #[error("Fail to send the Tx on the channel: {0}")]
+    PropagateTxError(#[from] Box<tokio::sync::mpsc::error::SendError<Transaction<TxValdiated>>>),
+    #[error("validation fail: {0}")]
+    ValidateError(String),
+    #[error("Tx asset fail to download because {0}")]
+    DownloadAssetError(String),
+    #[error("Save Tx error: {0}")]
+    SaveTxError(String),
+}
+
+pub type CallbackSender = oneshot::Sender<Result<(), EventProcessError>>;
+
+//Sending Tx interface.
+//Some marker type to define the sender source.
+pub struct RpcSender;
+#[derive(Clone)]
+pub struct P2pSender;
+pub struct TxResultSender;
+
+//use to send a tx to the event process.
+#[derive(Debug, Clone)]
+pub struct TxEventSender<T> {
+    sender: UnboundedSender<(Transaction<TxReceive>, Option<CallbackSender>)>,
+    _marker: PhantomData<T>,
+}
+
+//Manage send from the p2p source
+impl TxEventSender<P2pSender> {
+    pub fn build(
+        sender: UnboundedSender<(Transaction<TxReceive>, Option<CallbackSender>)>,
+    ) -> Self {
+        TxEventSender {
+            sender,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn send_tx(&self, tx: Transaction<TxCreate>) -> Result<(), EventProcessError> {
+        self.sender
+            .send((tx.into_received(TxReceive::P2P), None))
+            .map_err(|err| err.into())
+    }
+}
+
+//Manage send from the RPC source
+impl TxEventSender<RpcSender> {
+    pub fn build(
+        sender: UnboundedSender<(Transaction<TxReceive>, Option<CallbackSender>)>,
+    ) -> Self {
+        TxEventSender {
+            sender,
+            _marker: PhantomData,
+        }
+    }
+
+    pub async fn send_tx(&self, tx: Transaction<TxCreate>) -> Result<(), EventProcessError> {
+        let (sender, rx) = oneshot::channel();
+        self.sender
+            .send((tx.into_received(TxReceive::RPC), Some(sender)))
+            .map_err(EventProcessError::from)?;
+        rx.await?
+    }
+}
+
+//Manage send from the Tx result execution source
+impl TxEventSender<TxResultSender> {
+    pub fn build(
+        sender: UnboundedSender<(Transaction<TxReceive>, Option<CallbackSender>)>,
+    ) -> Self {
+        TxEventSender {
+            sender,
+            _marker: PhantomData,
+        }
+    }
+
+    pub async fn send_tx(&self, tx: Transaction<TxCreate>) -> Result<(), EventProcessError> {
+        let (sender, rx) = oneshot::channel();
+        self.sender
+            .send((tx.into_received(TxReceive::TXRESULT), Some(sender)))
+            .map_err(EventProcessError::from)?;
+        rx.await?
+    }
+}
+
+//Main event processing loog.
+pub async fn start_event_loop(
```

Comment:
```suggestion
pub async fn spawn_event_loop(
```

Maybe `spawn` instead of `start` to be explicit that this function will spawn a separate event loop "thread"? Alternatively maybe separate the creation of channels and then push the `tokio::spawn()` to higher level so that it's not "hidden" inside this function?

Reply:
I change the name.
For the spawn the method has 2 parts:  an init that create the channel, but it can change, and other thing can be added and the loop. We can split both by using a struct with an init() and loop() method, but it creates complexity just to avoid a hidden spawn. 
For me, the pb of hidden spawn is:
 * a looping task is created and should be monitored that why I return the Joinhandler.
 * it forces the runtime. If we need to change, the runtime will add it to the call.